internal/build/phases.go

```diff
@@ -15,17 +15,32 @@ const (
 	platformDir    = "/platform"
 )
 
-func (l *Lifecycle) Detect(ctx context.Context, networkMode string, volumes []string) error {
-	detect, err := l.NewPhase(
+type RunnerCleaner interface {
+	Cleanup() error
+	Run(ctx context.Context) error
+}
+
+type PhaseManager interface {
+	New(name string, ops ...PhaseOperation) (RunnerCleaner, error)
+	WithArgs(args ...string) PhaseOperation
```

Comment:
Ops shouldn't need to be created by a service class. This shouldn't be necessary and if I understand the PR it's only here for the purpose of testing the operations. If that true, I don't think this is ultimately the right use of this pattern.

Some alternatives could be a builder pattern or state verification on the end result by inspecting what mutations a concrete operation yields.

Reply:
I do not see how "state verification on the end result" would be possible given that the fields of `Phase` are unexported.

Could you elaborate on what a builder pattern might look like? I could see us creating something like a "phase configuration" that we pass to the phase manager and verify that the configuration is correct.

Reply:
There are two common builder patterns. Which one to pick is more or less per use case (for example if you wanted to pass around `BuildConfig`) and opinion:

```
// Single service constructs phase
PhaseBuilder()
 .withRoot(...)
 .withBind(..)
 .build() => Phase()

// A builder service and it's config
PhaseBuilder.build(BuildConfig().withRoot().withBind()) => Phase()
```

Reply:
A couple of thoughts:
- Could we just add getters for the unexported fields and avoid the larger refactor? I'd never be a fan of exporting just for the sake of testing, but adding read-only getters seems OK to me.

- If you didn't move the "opts" to their own methods, and instead left them as arguments on the `Lifecycle.NewPhase()` method, think you can still use `gomock` to inspect them. After all, functions are still first-class citizens. Did you try that and it didn't work?

Reply:
Update: I tested out bullet 2, and `gomock` either has a bug or doesn't allow something like this:

```go
mockLifecycle.EXPECT().NewPhase("build", []pack.LifecycleOp{build.WithArgs("x", "y", "z")})
mockLifecycle.NewPhase("build", build.WithArgs("x", "y", "z"))
```

It's strange because it __should__ work, per the test output, since functions are just pointers, but it still fails:

```
Expected call at /Users/ameyer/workspace/pack/internal/build/some_test.go:30 doesn't match the argument at index 1.
                Got: [0x14ac630]
                Want: is equal to [0x14ac630]
```

First, I tried moving the `WithArgs()` call to a stored variable, to ensure the reference/pointer was exactly the same, but there was no difference in the output above.

Reply:
(Also note, I added `NewPhase` to the interface to allow this, and added the `LifecycleOp` type alias for convenience)

Reply:
We stubbed out a possible implementation that we are somewhat happy with: https://github.com/buildpacks/pack/tree/di-phasemanager-phaseconfigprovider

The key concepts being here: https://github.com/buildpacks/pack/blob/e83f703002967a06ec82c5d51b1031ec0a230fea/internal/build/phases.go#L45

Thoughts?

Reply:
>```
> // Single service constructs phase
> PhaseBuilder()
>  .withRoot(...)
>  .withBind(..)
>  .build() => Phase()

I see how this would work but I don't see how this is testable, given the limitations discussed above. Am I missing something?

>```
> // A builder service and it's config
> PhaseBuilder.build(BuildConfig().withRoot().withBind()) => Phase()

This is approaching the implementation we are currently recommending. By inspecting the config object, we can build confidence that the phase is configured correctly.

Reply:
👍 Looks good. Achieves "state verification on the end result" in a cleaner way which I'm really happy with.